r/src/array_from_vector.cpp

```diff
@@ -1067,12 +1110,22 @@ std::shared_ptr<arrow::DataType> InferArrowTypeFromVector<VECSXP>(SEXP x) {
   if (Rf_inherits(x, "data.frame")) {
     return InferArrowTypeFromDataFrame(x);
   } else {
-    if (XLENGTH(x) == 0) {
-      Rcpp::stop(
-          "Requires at least one element to infer the values' type of a list vector");
-    }
+    SEXP ptype = Rf_getAttrib(x, symbols::ptype);
+    if (ptype == R_NilValue) {
+      if (XLENGTH(x) == 0) {
+        Rcpp::stop(
+            "Requires at least one element to infer the values' type of a list vector");
+      }
 
-    return arrow::list(InferArrowType(VECTOR_ELT(x, 0)));
+      return arrow::list(InferArrowType(VECTOR_ELT(x, 0)));
+    } else {
+      // special case list(raw()) -> BinaryArray
+      if (TYPEOF(ptype) == RAWSXP) {
+        return arrow::binary();
+      }
+
+      return arrow::list(InferArrowType(ptype));
```

Comment:
Oh yeah that makes sense. Although this only looks at the attribute, not specifically that it is a `vctrs_list_of` but on the arrow -> R conversion, I think it does not hurt to make it a `vctrs_list_of`

Reply:
Done. I had to modify the roundtrip checks to use `expect_equivalent()` because a roundtrip might add information: 

```
list() -> List Array -> list_of( ptype = <from list array value type>) 
```